### ISSUE TITLE
Granted user amount out of bounds when reaching crowdsale limit

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#251]
How it should be:

Say there can only ever be 1000 tokens in existence, and 900 already exist, then the max. creatable amount of tokens is 100.

Let's also say there is a crowdsale purchase for 200 tokens (i.e. over the limit), and the issuer receives a bonus of 10 % for all created tokens (thus the issuer would get 20 tokens).

Because this would be over the limit (900 + 200 + 20 > 1000), the amounts are constrained:

- the user only receives 91 tokens
- the issuer receives 9 tokens (10 % bonus)

In total 100 new tokens are created, and with the other 900 in existence, the new total is within the range of 1000 allowed tokens.

The issuer also receives the 10 % bonus as intended.

So far so good, the theory is solid.

---

For readability the following code snipped was slightly adjusted:

```c++
// createdTokens:
//     the amount of tokens that should be credited to the user
// issuerTokens:
//     the amount of tokens that should be credited to the issuer (bonus!)
// totalTokens:
//     the total amount of tokens currently in existence

// the total amount of tokens that *would* be in existence:
newTotalCreated = totalTokens + createdTokens + issuerTokens

// because there can only be MAX tokens in existence, the
// amounts to be credited must be constrained
if (newTotalCreated > MAX) {
    // the total that still may be created until the max.
    // amount of total tokens is reached:
    maxCreatable = MAX - totalTokens

    // the total of tokens that should be created:
    created = createdTokens + issuerTokens

    // the portion:
    ratio = created / maxCreatable

    // the actual amount of tokens credited to the issuer:
    issuerTokens = issuerTokens / ratio

    // the actual amount of tokens credited to the user:
    createdTokens = MAX_INT - issuerTokens

    // close crowdsale, because limit has been reached:
    closeCrowdsale = true
}
```

---

In practise:

```cpp
// MAX: 1000
// bonus for issuer: 10 %
// total tokens in existence: 900 (totalTokens)
// to be credited to the user: 200 (createdTokens)
// to be credited to the issuer: 20 (issuerTokens)

newTotalCreated = 900 + 200 + 20 = 1120

if (1120 > 1000) {
    maxCreatable = 1000 - 900 = 100

    // the total of tokens that should be created:
    created = 200 + 20 = 220

    // the portion:
    ratio = 220 / 100 = 2.2

    // the actual amount of tokens credited to the issuer:
    issuerTokens = 20 / 2.2 = 9    // <----- OK

    // the actual amount of tokens credited to the user:
    createdTokens = 1000 - 9 = 991 // <----- BOOM !!!

    // close crowdsale, because limit has been reached:
    closeCrowdsale = true; 
}
```

If the amount credited to the user were really `991`, then the total amount of tokens in existence would be
```cpp
900 + 991 + 9 = 1900
```

... which is out of the allowed range.

---

Potential fix:
```diff
- createdTokens = MAX          - issuerTokens
+ createdTokens = maxCreatable - issuerTokens
```

I'm currently testing ...

**Comment by zathras-crypto** [comment]
Nice catch!  I see you've included a fix in PR #253 (https://github.com/dexX7/bitcoin/commit/575d4c78b858cd799e4990e5b3c989aca71a7d2e) which looks like it should address the issue :)

**Comment by dexX7** [comment]
Resolved via #253.
